File: online-training-platform/leetcode/278. First Bad Version.cpp
// The API isBadVersion is defined for you.
// bool isBadVersion(int version);

class Solution {
public:
    
    int bs(int min, int max){
        long long mid;
        bool t1,t2;
        while( max - min > 1){
            mid = min + (max-min)/2;
            t1 = isBadVersion(mid);
            t2 = isBadVersion(mid+1);
            if(!t1 && t2) return mid+1;
            if(!t1) min = mid;
            else max = mid;
            
        }
        t1 = isBadVersion(min);
        t2 = isBadVersion(max);
        
        return (!t1 && t2)? max : min;
    }
    
    int firstBadVersion(int n) {
        int r = bs(1,n);
        return r;
    }
};

File: online-training-platform/leetcode/278. First Bad Version.kt
/* The isBadVersion API is defined in the parent class VersionControl.
      fun isBadVersion(version: Int) : Boolean {} */

      class Solution: VersionControl() {
        override fun firstBadVersion(n: Int) : Int {
            var low = 1; var high = n
            while(low < high) {
                val mid = low + (high - low) / 2
                if(isBadVersion(mid).not()) {
                    low = mid + 1
                } else {
                    high = mid
                }
            }
    
            return low
        }
    }